backend/cmm_helpers.ml

```diff
@@ -2526,172 +2566,202 @@ let tuplify_function arity =
-          get_field_gen Asttypes.Mutable (Cvar clos) 4 (dbg ()),
-          curry_fun
-            (get_field_gen Asttypes.Mutable (Cvar clos) 3 (dbg ()) :: args)
-            newclos (n - 1) )
-  in
+  let narity = List.length arity in
   let fun_name =
-    "caml_curry" ^ Int.to_string arity
-    ^ (if nlocal > 0 then "L" ^ Int.to_string nlocal else "")
+    curry_function_sym (Lambda.Curried { nlocal }) arity result
     ^ "_"
-    ^ Int.to_string (arity - 1)
+    ^ Int.to_string (narity - 1)
   in
+  let args_type = List.rev arity in
   let fun_dbg = placeholder_fun_dbg ~human_name:fun_name in
   Cfunction
     { fun_name;
-      fun_args = [VP.create last_arg, typ_val; VP.create last_clos, typ_val];
-      fun_body = curry_fun [] last_clos (arity - 1);
+      fun_args =
+        [VP.create last_arg, List.hd args_type; VP.create last_clos, typ_val];
+      fun_body =
+        make_curry_apply result narity (List.tl args_type) [Cvar last_arg]
+          last_clos (narity - 1);
       fun_codegen_options = [];
       fun_dbg;
       fun_poll = Default_poll
     }
 
-let rec intermediate_curry_functions ~nlocal ~arity num =
+let intermediate_curry_functions nlocal arity result =
+  let name1 = curry_function_sym (Lambda.Curried { nlocal }) arity result in
+  let narity = List.length arity in
   let dbg = placeholder_dbg in
-  if num = arity - 1
-  then [final_curry_function ~nlocal ~arity]
-  else
-    let name1 =
-      "caml_curry" ^ Int.to_string arity
-      ^ if nlocal > 0 then "L" ^ Int.to_string nlocal else ""
-    in
-    let name2 = if num = 0 then name1 else name1 ^ "_" ^ Int.to_string num in
-    let arg = V.create_local "arg" and clos = V.create_local "clos" in
-    let fun_dbg = placeholder_fun_dbg ~human_name:name2 in
-    let mode : Lambda.alloc_mode =
-      if num >= arity - nlocal then Lambda.alloc_local else Lambda.alloc_heap
-    in
-    let curried n : Clambda.arity = Curried { nlocal = min nlocal n }, n in
-    Cfunction
-      { fun_name = name2;
-        fun_args = [VP.create arg, typ_val; VP.create clos, typ_val];
-        fun_body =
-          (if arity - num > 2 && arity <= max_arity_optimized
-          then
+  let rec loop accumulated_args remaining_args num =
+    match remaining_args with
+    | [] -> Misc.fatal_error "Empty arity for [intermediate_curry_functions]"
+    | [_] -> [final_curry_function nlocal arity result]
+    | arg_type :: remaining_args ->
+      let name2 = if num = 0 then name1 else name1 ^ "_" ^ Int.to_string num in
+      let arg = V.create_local "arg" and clos = V.create_local "clos" in
+      let fun_dbg = placeholder_fun_dbg ~human_name:name2 in
+      let mode : Lambda.alloc_mode =
+        if num >= narity - nlocal then Lambda.alloc_local else Lambda.alloc_heap
+      in
+      let curried n = Lambda.Curried { nlocal = min nlocal n }, n in
+      let has_nary = curry_clos_has_nary_application ~narity (num + 1) in
+      let header_size = if has_nary then 3 else 2 in
```

Comment:
header_size seems like a bad name (this isn't related to the header of the block), maybe function_slot_size?